src/types.cpp

```diff
@@ -355,13 +356,26 @@ SizedType CreatePointer(const SizedType &pointee_type, AddrSpace as)
   return ty;
 }
 
-SizedType CreateRecord(size_t size, const std::string &name)
+SizedType CreateRecord(const std::string &name, size_t size, BPFtrace &bpftrace)
 {
-  auto ty = SizedType(Type::record, size);
+  auto record = bpftrace.structs_.insert(
```

Comment:
I don't think we should tie the typesystem to BPFtrace like this. A type is just a
description so we know how to deal with it. Instance of the type (all the structs coming
from btf/clang) are not relevant to the type system, they should be dealt with at another level.

This also makes #1845 harder. Right now we can use the type system and sill keep it simple, if we add BPFtrace in you have
to pull in all the stuff defined in BPFtrace (i.e. everything we have) which makes it way harder to use as lib.

This structmap type should really only exists between the clang,btf and semantic phases. clang,btf define structs and store
them in the map. The semantic phase finds the types it need, which it knows from either pointer cast or btf,  and assigns it
to ast nodes. All unused types can be discarded at that point. 
Keeping it in BPFtrace is easier though, a single object that keeps all the state instead of having to pass state between phases.

Having a simple abstraction to access this instead of dealing with bpftrace_.structs_ directly makes sense. Maybe a `StructManager`
similar to `MapManager`. But that doesn't belong in the typesystem.

Not needing bpftrace for all the type related stuff should also keep the btf & clang code simpler, don't have to keep passing bpftrace around.

Reply:
Hm, this sounds reasonable. If I understand correctly, you're suggesting to keep the `bpftrace.structs_` map which will be populated from ClangParser (the way it is done now) and its entries will be copied into AST nodes during semantic analysis (perhaps using an abstraction layer). In other words, instead of holding a pointer to the struct map, `SizedType` will contain a `Struct` definition by value.

This will introduce some struct copying, which I wanted to avoid, but I understand that having the type system independent from BPFtrace is more important.

Reply:
A lot of sizedtype handling is currently by value anyway. It would be nice if types are immutable so we only have to store on copy and avoid corrupting it, but thats a different issue.

I think we can have a simple `StructManager` class that handles add /lookup.

So in clang_parser you can write something like:

1. bpftrace.structs.Add(name)
1. bpftrace.structs.AddField(name, field...)
1. bpftrace.structs.Lookup(name)

1,2 would be relevant to the clang/btf phase for populating types.
3. would be used in the semantic phase to assign types to ast nodes

I *think* that should work ok. Its not that different from what you have now. It would just move the logic from CreateRecord to a wrapper class.

Reply:
`StructManager` sounds good, it shouldn't be a problem to implement it.

What will be a problem is storing full field definitions inside `SizedType` by value. There are 2 issues:

1. How do we handle fields that are pointers to structs? For pointers, we store the pointed type inside, however, this will be likely impossible with recursive types.
2. There are some struct types that are created before their definition appears in the struct map. These are in particular (1) kfunc args types (created from `FieldAnalyser`) and (2) types of struct fields that are pointers to forward-declared structs (created in `ClangParser`). 
 
I'm afraid that these will prevent us from storing full field definitions inside `SizedType`. We could solve no. 2 by updating the relevant types after all struct types definitions are retrieved, but no.1 will be problematic unless we somehow rework the way pointer types are stored.

Or we leave `SizedType` to contain a reference to the struct map (since it's a `shared_ptr`, it should survive even if the entry in the struct map is deleted later) and just move the logic of adding entries to the struct map from `CreateRecord` to `StructManager`.

Reply:
Ah yeah recursive types are a pain. But as long as every type is unique (e.g.  every int8 points to the same immutable  type) that shouldn't be an issue. All we'd need is a way to forward declare structs and  extend them later. As everything is pointing to the same Type object we wouldn't have to do anything later. When the  fields of `Struct Y` get filled `Struct S { Y y }` will be updated automatically as everything references the same Y. But if we do that we do need some kind of type storage, not just for structs but for all types.

Thats also how LLVM handles its types, every type is unique within the context. Basic types are pre-allocated, https://llvm.org/doxygen/classllvm_1_1LLVMContextImpl.html  , https://llvm.org/doxygen/classllvm_1_1Type.html#aa75984a442f2379de0c66018201fa628  and structs must be unique

> There are two different kinds of struct types: Literal structs and Identified structs.

> Literal struct types (e.g. { i32, i32 }) are uniqued structurally, and must always have a body when created. You can get one of these by using one of the StructType::get() forms.

> Identified structs (e.g. foo or %42) may optionally have a name and are not uniqued. The names for identified structs are managed at the LLVMContext level, so there can only be a single identified struct with a given name in a particular LLVMContext. Identified structs may also optionally be opaque (have no body specified). You get one of these by using one of the StructType::create() forms.

I think that would make sense for us too. No more type copying, only assignment. Getting there will be quite some work but maybe this PR can be a first step. Then we can leave it as is and extend the type storage later. If we keep the type storage as a separate type we don't need to depend on bpftrace.h, but that's easy to refactor later.

Reply:
Actually, this is pretty much how it is implemented right now (in fact, it's inspired by the LLVM type system). All struct types are unique and stored in `bpftrace.struct_map`. `SizedType` then contains just a reference to the unique type (and thanks to that, the actual definitions can be forward-declared and populated later).

So the only improvement of the current implementation will be to use a `StructManager` instead of passing `BPFtrace` to `CreateRecord`.

Reply:
Yeah, but doing it for all types and making sure no copies happen (so we can drop the copy constructor) is a bit more work. 


 > So the only improvement of the current implementation will be to use a StructManager instead of passing BPFtrace to CreateRecord.

Yes, sounds good

Reply:
> Yeah, but doing it for all types and making sure no copies happen (so we can drop the copy constructor) is a bit more work.

Of course, hopefully, we'll be able to build on this approach (e.g. eventually transforming `StructManager` into `TypeManager`).

Anyway, I implemented the discussed changes, PR should be ready for another review.